## Importing Libraries

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import classification_report,accuracy_score,confusion_matrix
from sklearn.model_selection import StratifiedKFold,cross_val_score,cross_val_predict
from sklearn.feature_selection import SelectKBest,chi2,f_classif
from sklearn.model_selection import train_test_split,GridSearchCV,RandomizedSearchCV
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier,VotingClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier

In [4]:
import warnings
warnings.filterwarnings("ignore")

In [5]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [6]:
df=pd.read_csv(r'C:\Users\omdes\Desktop\data\teks\All_data_sets\Minsk2020_ALS_dataset[1].csv')

In [5]:
df.head()

ID Sex  Age      J1_a      J3_a      J5_a     J55_a      S1_a      S3_a  \
0   8   M   58  0.321817  0.141230  0.199128  0.923634  6.044559  3.196477   
1  20   F   57  0.344026  0.177032  0.206458  0.827714  1.967728  0.856639   
2  21   F   58  0.264740  0.148228  0.177078  0.532566  1.850893  0.942743   
3  22   F   70  0.455793  0.174870  0.243660  0.962641  2.883768  1.284926   
4  24   M   66  0.269335  0.143961  0.167465  0.547745  2.327924  1.164109   

       S5_a     S11_a      S55_a      DPF_a     PFR_a     PPE_a     PVI_a  \
0  3.770575  5.709480  10.080498  58.483755  0.261201  0.953932  0.497905   
1  1.179851  2.050048   5.058724  68.163593  0.214815  0.917383  0.496096   
2  1.071950  1.502212   2.953059  60.971524  0.250832  0.697039  0.484256   
3  1.915058  2.929953   4.261813  51.304348  0.394053  1.400853  0.710780   
4  1.420891  2.141512   3.741182  63.005780  0.280716  0.604214  0.464793   

       HNR_a  GNEa_{\mu}  GNEa_{\sigma}  Ha(1)_{mu}  Ha(2)_{mu}  Ha(3)_{mu}  \
0   5.611603    0.916190       0.033908  -30.714386  -16.287423  -21.036437   
1  10.430052    0.891161       0.033989   -5.771198   -8.339282  -13.213540   
2  20.946624    0.949725       0.025703   -3.989259  -11.210482  -13.362496   
3   3.902827    0.941889       0.025321   -8.962008   -4.678120   -9.086231   
4  21.511197    0.899420       0.033984   -4.946396   -6.273316   -1.885844   

   Ha(4)_{mu}  Ha(5)_{mu}  Ha(6)_{mu}  Ha(7)_{mu}  Ha(8)_{mu}  Ha(1)_{sd}  \
0  -10.515995  -18.079135  -22.050308  -27.556129  -32.988570    4.137014   
1   -6.001741  -16.789944  -27.686578  -33.427112  -38.601498    1.081714   
2  -14.044139  -20.689930  -17.478403  -17.045505  -30.253636    0.971833   
3  -11.007827  -11.892048  -17.322011  -15.533825  -22.283258    1.478457   
4  -10.701212  -23.082131  -23.639773  -17.307442  -27.578969    1.381583   

   Ha(2)_{sd}  Ha(3)_{sd}  Ha(4)_{sd}  Ha(5)_{sd}  Ha(6)_{sd}  Ha(7)_{sd}  \
0    5.378604    3.296739    7.365126    5.773458    9.784643    3.718647   
1    2.295224    2.235447    2.639255    2.513208    5.371963    6.869505   
2    1.808590    2.984416    3.812197    6.729836    4.897549    6.487032   
3    2.039120    1.179530    1.709350    2.107210    2.479865    2.293785   
4    1.287978    0.927360    2.718745    8.362299    4.476388    3.365784   

   Ha(8)_{sd}  Ha(1)_{rel}  Ha(2)_{rel}  Ha(3)_{rel}  Ha(4)_{rel}  \
0    7.867399     0.028693     0.046155     0.041096     0.055925   
1    5.114703     0.145923     0.094034     0.064729     0.115727   
2    7.651709     0.201569     0.076810     0.061174     0.056003   
3    3.408103     0.095781     0.148871     0.097411     0.078634   
4    3.314668     0.158028     0.132252     0.355467     0.074516   

   Ha(5)_{rel}  Ha(6)_{rel}  Ha(7)_{rel}  Ha(8)_{rel}    CCa(1)     CCa(2)  \
0     0.041924     0.031412     0.031975     0.024476 -7.456684  -8.269332   
1     0.051805     0.030249     0.024816     0.022875  7.588634  -2.869248   
2     0.036470     0.044691     0.042494     0.026382  0.883309 -14.569909   
3     0.071432     0.050500     0.056093     0.038924 -2.578846  -9.112154   
4     0.031802     0.035567     0.048372     0.032369  3.819291  -8.518077   

      CCa(3)     CCa(4)     CCa(5)     CCa(6)    CCa(7)     CCa(8)     CCa(9)  \
0 -12.385919 -11.454670  -8.556208   1.708060 -0.324135 -18.335941   6.885555   
1 -13.641117  -3.802710  -3.273136   3.394290 -4.651991  -9.201281  -8.219403   
2  -2.596838  -3.512899 -15.110428  -4.884618 -8.427400 -17.562101  -7.415726   
3  -0.259772   0.411847 -20.882510  10.407598 -8.844432  -7.785514   0.207478   
4  -8.279661  -8.450896  -9.403074  -4.226639 -0.531899  -0.458887 -14.646083   

     CCa(10)   CCa(11)   CCa(12)   dCCa(1)   dCCa(2)   dCCa(3)   dCCa(4)  \
0  -8.679375 -3.506784  1.763508 -0.002462  0.013113  0.008769 -0.012167   
1  -8.627542  1.379169 -0.987091 -0.000023 -0.005384  0.003537  0.008648   
2  -9.102621 -9.543975 -9.479419  0.008926  0.004305 -0.002191  0.00225

## Label Encoding

In [7]:
col_to_encode=df.select_dtypes(include='object').columns

In [8]:
le=LabelEncoder()
for i in col_to_encode:
    df[i]=le.fit_transform(df[i])

## checking the initial score before SelectKBest

In [9]:
x=df.iloc[:,:-1]
y=df['Diagnosis (ALS)']

In [9]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [10]:
lr=LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

In [12]:
scores=cross_val_score(estimator= lr,X=x,y=y,cv=StratifiedKFold(5))
print(scores.mean())

0.7


In [13]:
y_pred=lr.predict(x_test)

In [14]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.75      0.86      0.80         7
           1       0.80      0.67      0.73         6

    accuracy                           0.77        13
   macro avg       0.78      0.76      0.76        13
weighted avg       0.77      0.77      0.77        13



## Using SelectKBest

In [15]:
col=x.shape[1]

In [16]:
rf=RandomForestClassifier()
lr=LinearRegression()
scores_dic={}
for i in range(1,col):
    kbest=SelectKBest(f_classif,k=i)
    x_selected=kbest.fit_transform(x,y)
    scores = cross_val_score(estimator= rf,X=x_selected,y=y,cv=StratifiedKFold(5))
    scores_dic[i]=scores.mean()

In [27]:
top_3_scores = dict(sorted(scores_dic.items(), key=lambda item: item[1], reverse=True)[:3])

print("Top 3 accuracy scores:")
for k, v in top_3_scores.items():
    print(f"Number of features: {k}, Accuracy: {v}")
#RandomForest

Top 3 accuracy scores:
Number of features: 12, Accuracy: 0.8589743589743589
Number of features: 19, Accuracy: 0.8448717948717951
Number of features: 15, Accuracy: 0.8448717948717949


In [27]:
top_3_scores = dict(sorted(scores_dic.items(), key=lambda item: item[1], reverse=True)[:3])

print("Top 3 accuracy scores:")
for k, v in top_3_scores.items():
    print(f"Number of features: {k}, Accuracy: {v}")
#LogisticRegression

Top 3 accuracy scores:
Number of features: 4, Accuracy: 0.7961538461538462
Number of features: 12, Accuracy: 0.7653846153846153
Number of features: 3, Accuracy: 0.764102564102564


#### selected k=12 as it is giving best accuracy in random Forest and 2nd highest in Logistic regression

In [10]:
kbest=SelectKBest(f_classif,k=12)
kbest.fit_transform(x,y)

array([[ 9.23634406e-01,  2.61201275e-01,  9.53931647e-01,
         4.97904860e-01,  3.19746497e-02,  5.48737410e-01,
         6.83156779e+00,  2.95775708e-02,  5.29067088e+00,
        -7.16254751e+00,  4.82547595e+00,  8.33498083e+02],
       [ 8.27714472e-01,  2.14815072e-01,  9.17383049e-01,
         4.96095925e-01,  2.48159789e-02,  7.53773055e-01,
         1.96563010e+01,  8.45142365e-02,  8.01368362e+00,
         6.70750243e+00,  5.72932234e+00,  5.61802625e+02],
       [ 5.32565829e-01,  2.50831736e-01,  6.97038661e-01,
         4.84256028e-01,  4.24943560e-02,  5.13454294e-01,
         2.62470321e+01,  1.90497985e-01,  5.39853075e+00,
        -1.59782411e+01,  8.25848843e+00,  7.96723440e+02],
       [ 9.62640787e-01,  3.94052825e-01,  1.40085272e+00,
         7.10780139e-01,  5.60927669e-02,  5.31370407e-01,
         1.44974271e+01,  4.41358562e-02, -2.03201060e+00,
         7.38321172e+00,  5.44713714e+00,  3.59409974e+02],
       [ 5.47744920e-01,  2.80716281e-01,  6.0421385

In [11]:
kbest.get_support()

array([False, False, False, False, False, False,  True, False, False,
       False, False, False, False,  True,  True,  True, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True,  True,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False, False, False, False,
        True, False, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [12]:
x_best_cols=x.columns[kbest.get_support()]

In [13]:
x_best_cols

Index(['J55_a', 'PFR_a', 'PPE_a', 'PVI_a', 'Ha(7)_{rel}', 'PVI_i', 'HNR_i',
       'Hi(1)_{rel}', 'CCi(2)', 'CCi(6)', 'd_1', 'F2_{conv}'],
      dtype='object')

In [14]:
x_best=df[x_best_cols]

In [15]:
x_train,x_test,y_train,y_test=train_test_split(x_best,y,test_size=0.2,random_state=42)

In [16]:
y.value_counts()

Diagnosis (ALS)
0    33
1    31
Name: count, dtype: int64

#### not using Resampling techniques as the data is very slightly imbalanced

## Model Buliding

In [58]:
models=[LogisticRegression(),DecisionTreeClassifier(),RandomForestClassifier(),GaussianNB(),KNeighborsClassifier(),SVC(),XGBClassifier(),AdaBoostClassifier(estimator=DecisionTreeClassifier(max_depth=1),n_estimators=50)]

In [59]:
for i in models:
    y_pred=cross_val_predict(estimator=i,X=x_best,y=y,cv=StratifiedKFold(5),n_jobs=-1)
    print(i)
    print('Confusion matrix:')
    print(confusion_matrix(y,y_pred))
    print('_ _ _ '*5)
    print('Classification Report:')
    print(classification_report(y,y_pred))
    print('_ _ _ '*15)

LogisticRegression()
Confusion matrix:
[[29  4]
 [11 20]]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.88      0.79        33
           1       0.83      0.65      0.73        31

    accuracy                           0.77        64
   macro avg       0.78      0.76      0.76        64
weighted avg       0.78      0.77      0.76        64

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
DecisionTreeClassifier()
Confusion matrix:
[[28  5]
 [ 9 22]]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Classification Report:
              precision    recall  f1-score   support

           0       0.76      0.85      0.80        33
           1       0.81      0.71      0.76        31

    accuracy                           0.78        64
   macro avg       0.79      0.78      0.78        64
weighted avg       0.78      0.78      0.78        64

_ _ _ _ _ _ _ _ _ _ _ _ _ _

### Selecting the top 3 algorithms and applying GridSearchCV
- random Forest
- Namve bayes
- XG Boost 

#### random Forest

In [22]:
rfc_grid={
    'criterion':['gini','entropy'],
    'n_estimators':[100,150,200],
    'max_depth':[None,10,12,15],
    'min_samples_split':[2,5,10],
    'min_samples_leaf':[1,2,4],
}

In [23]:
rfgscv=GridSearchCV(estimator=RandomForestClassifier(),n_jobs=1,cv=StratifiedKFold(5),verbose=2,param_grid=rfc_grid)

In [24]:
rfgscv.fit(X=x_best,y=y)

Fitting 5 folds for each of 216 candidates, totalling 1080 fits
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.3s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.3s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.4s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.4s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.3s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=150; tot

[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=5, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   0.6s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   0.6s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   0.7s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   0.6s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   0.6s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=None, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=None, min_s

[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.5s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.6s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=1, min_s

[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   0.6s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=5, n_estimators=200; total time=   0.5s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.3s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.3s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=150; total time=   0.4s
[CV] END criterion=gini, max_depth=10, min_samples_leaf=2,

[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.1s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=2, n_estimators=150; total time=   0.1s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.2s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.2s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.2s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.5s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.4s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=1, min_s

[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=100; total time=   0.0s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=150; total time=   0.1s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=150; total time=   0.1s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=150; total time=   0.2s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=150; total time=   0.3s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=150; total time=   0.3s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.5s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.2s
[CV] END criterion=gini, max_depth=12, min_samples_leaf=

[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   0.5s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.1s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.0s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.0s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.3s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.2s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=1, min_s

[CV] END criterion=gini, max_depth=15, min_samples_leaf=2, min_samples_split=10, n_estimators=150; total time=   0.3s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.5s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.3s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.3s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.3s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.3s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.2s
[CV] END criterion=gini, max_depth=15, min_samples_leaf=4,

[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.3s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.3s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.3s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.4s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.2s
[CV] END criterion=entro

[CV] END criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.3s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.1s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.2s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.2s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=None, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=e

[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.2s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.5s
[CV] END criterion=entropy, max_depth=10

[CV] END criterion=entropy, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.1s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=2, min_samples_split=10, n_estimators=200; total time=   0.1s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=10, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=

[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.2s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.3s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.4s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.4s
[CV] END criterion=entropy, max_depth=12

[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.2s
[CV] END criterion=entropy, max_depth=12, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.3s
[CV] END criterion=entropy, max_depth=12

[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.4s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.4s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.2s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.2s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   0.1s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=10, n_estimators=100; total time=   0.1s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=10, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=1, min_samples_split=10, n_estimators=100; total time=   0.0s
[CV] END criterion=entropy, max_depth

[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=100; total time=   0.2s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.3s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.3s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.3s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=150; total time=   0.1s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=200; total time=   0.1s
[CV] END criterion=entropy, max_depth=15, min_samples_leaf=4, min_samples_split=2, n_estimators=200; total time=   0.1s
[CV] END criterion=entropy, max_depth=15

GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=None, shuffle=False),
             estimator=RandomForestClassifier(), n_jobs=1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [None, 10, 12, 15],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 150, 200]},
             verbose=2)

In [29]:
rfgscv.best_score_

0.8435897435897436

In [27]:
best_rfc = rfgscv.best_estimator_

In [28]:
y_pred_voting=cross_val_predict(estimator=best_rfc,X=x_best,y=y,cv=StratifiedKFold(5),n_jobs=-1)
print('Confusion Matrix:\n', confusion_matrix(y, y_pred_voting))
print('_ _ _ ' * 5)
print('Classification Report:\n', classification_report(y, y_pred_voting))

Confusion Matrix:
 [[29  4]
 [ 9 22]]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Classification Report:
               precision    recall  f1-score   support

           0       0.76      0.88      0.82        33
           1       0.85      0.71      0.77        31

    accuracy                           0.80        64
   macro avg       0.80      0.79      0.79        64
weighted avg       0.80      0.80      0.80        64



#### Naive bayes

In [30]:
nb_grid = {
    'var_smoothing': [1e-9, 1e-8, 1e-7, 1e-6, 1e-5, 1e-4, 1e-3, 1e-2, 0.1, 1.0]
}


In [31]:
nbgscv=GridSearchCV(estimator=GaussianNB(),n_jobs=-1,verbose=2,param_grid=nb_grid,cv=StratifiedKFold(5))

In [32]:
nbgscv.fit(X=x_best,y=y)

Fitting 5 folds for each of 10 candidates, totalling 50 fits


GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=None, shuffle=False),
             estimator=GaussianNB(), n_jobs=-1,
             param_grid={'var_smoothing': [1e-09, 1e-08, 1e-07, 1e-06, 1e-05,
                                           0.0001, 0.001, 0.01, 0.1, 1.0]},
             verbose=2)

In [33]:
nbgscv.best_score_

0.8128205128205129

In [34]:
best_nbc = nbgscv.best_estimator_

In [35]:
y_pred_voting=cross_val_predict(estimator=best_nbc,X=x_best,y=y,cv=StratifiedKFold(5),n_jobs=-1)
print('Confusion Matrix:\n', confusion_matrix(y, y_pred_voting))
print('_ _ _ ' * 5)
print('Classification Report:\n', classification_report(y, y_pred_voting))

Confusion Matrix:
 [[32  1]
 [11 20]]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Classification Report:
               precision    recall  f1-score   support

           0       0.74      0.97      0.84        33
           1       0.95      0.65      0.77        31

    accuracy                           0.81        64
   macro avg       0.85      0.81      0.81        64
weighted avg       0.85      0.81      0.81        64



#### xgboost

In [36]:
xgb_grid = {
    'n_estimators': [50, 100, 200, 300],
    'max_depth': [3, 6, 10, 12],
    'learning_rate': [0.01, 0.05, 0.1, 0.2],
    'subsample': [0.8, 1.0],
    'colsample_bytree': [0.8, 1.0],
    'gamma': [0, 0.1, 0.2],
    'reg_alpha': [0, 0.01, 0.1],
    'reg_lambda': [0.01, 0.1, 1.0]
}

In [37]:
xgbgscv = RandomizedSearchCV(estimator=XGBClassifier(),param_distributions=xgb_grid,n_iter=100,cv=StratifiedKFold(5),verbose=2,n_jobs=-1,)

In [38]:
xgbgscv.fit(X=x_best,y=y)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


RandomizedSearchCV(cv=StratifiedKFold(n_splits=5, random_state=None, shuffle=False),
                   estimator=XGBClassifier(base_score=None, booster=None,
                                           callbacks=None,
                                           colsample_bylevel=None,
                                           colsample_bynode=None,
                                           colsample_bytree=None, device=None,
                                           early_stopping_rounds=None,
                                           enable_categorical=False,
                                           eval_metric=None, feature_types=None,
                                           gamma=None, grow_policy=None,
                                           impor...
                                           multi_strategy=None,
                                           n_estimators=None, n_jobs=None,
                                           num_parallel_tree=None,
                                           random_state=None, ...),
                   n_iter=100, n_jobs=-1,
                   param_distributions={'colsample_bytree': [0.8, 1.0],
                                        'gamma': [0, 0.1, 0.2],
                                        'learning_rate': [0.01, 0.05, 0.1, 0.2],
                                        'max_depth': [3, 6, 10, 12],
                                        'n_estimators': [50, 100, 200, 300],
                                        'reg_alpha': [0, 0.01, 0.1],
                                        'reg_lambda': [0.01, 0.1, 1.0],
                                        'subsample': [0.8, 1.0]},
                   verbose=2)

In [39]:
xgbgscv.best_score_

0.8435897435897436

In [40]:
best_xgbc = xgbgscv.best_estimator_

In [41]:
y_pred_voting=cross_val_predict(estimator=best_xgbc,X=x_best,y=y,cv=StratifiedKFold(5),n_jobs=-1)
print('Confusion Matrix:\n', confusion_matrix(y, y_pred_voting))
print('_ _ _ ' * 5)
print('Classification Report:\n', classification_report(y, y_pred_voting))

Confusion Matrix:
 [[31  2]
 [ 8 23]]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Classification Report:
               precision    recall  f1-score   support

           0       0.79      0.94      0.86        33
           1       0.92      0.74      0.82        31

    accuracy                           0.84        64
   macro avg       0.86      0.84      0.84        64
weighted avg       0.86      0.84      0.84        64



####  voting classifier

In [110]:
voting_clf=VotingClassifier(
    estimators=[
        ('XGBoost',best_xgbc),
        ('RandomForest',best_rfc),
#         ('GaussianNB',best_nbc)
    ]
    ,voting='hard')

In [109]:
y_pred_voting=cross_val_predict(estimator=voting_clf,X=x_best,y=y,cv=StratifiedKFold(5),n_jobs=-1)
print('Confusion Matrix:\n', confusion_matrix(y, y_pred_voting))
print('_ _ _ ' * 5)
print('Classification Report:\n', classification_report(y, y_pred_voting))

Confusion Matrix:
 [[30  3]
 [ 7 24]]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Classification Report:
               precision    recall  f1-score   support

           0       0.81      0.91      0.86        33
           1       0.89      0.77      0.83        31

    accuracy                           0.84        64
   macro avg       0.85      0.84      0.84        64
weighted avg       0.85      0.84      0.84        64



###### 

## conclusion: 
XGBoost classifier is giving the best accuracy of 84.3% . <br>
Confusion Matrix:<br>
 [[31  2]  <br>
 [ 8 23]]  <br>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Classification Report: <br>

                   precision   recall   f1-score   support
      
          0          0.79      0.94      0.86        33 
          1          0.92      0.74      0.82        31